File: Algo_Problems/Sorting/quickSort.cpp
#include<bits/stdc++.h>
using namespace std;

// Key -> Select a pivot element and keep smaller elements before the pivot element and greater elements
// after the pivot element. Then sort the smaller and greater elements again.

int partition(int a[], int l, int r){
    int pivot = a[r];
    int i = l-1;
    for(int j=l;j<r;j++){
        if(a[j]<pivot){
            i++;
            swap(a[i], a[j]);
        }
    }
    swap(a[i+1], a[r]);
    cout<<i+1<<"\t";
    return i+1;
}

void quickSort(int a[], int l, int r){
    if(l<r){
        int pi = partition(a, l, r);
        quickSort(a, l, pi-1);
        quickSort(a, pi+1, r);
    }
}

int main(){
    int a[] = {4,2,9,6,1,3};
    quickSort(a, 0, 5);
    cout<<endl;
    for(int i=0;i<6;i++){
        cout<<a[i]<<"  ";
    }
}

File: Algo_Problems/Sorting/QuickSort.java
package Algo_Problems.Sorting;

/*
 * Key -> Select a pivot element and keep smaller elements before the pivot element and greater elements
 *after the pivot element. Then sort the smaller and greater elements again.
*/

public class QuickSort {
    public static void main(String[] args) {
        int nums[] = { 5, 2, 10, 9, 1, 4, 3, 7 };
        solve(nums);
        for (int i = 0; i < nums.length; i++) {
            System.out.print(nums[i] + " ");
        }
        System.out.println("");
    }

    static void solve(int nums[]) {
        quickSort(nums, 0, nums.length - 1);
    }

    static void quickSort(int nums[], int l, int r) {
        if (l < r) {
            int pivotIdx = partition(nums, l, r);
            quickSort(nums, l, pivotIdx - 1);
            quickSort(nums, pivotIdx + 1, r);
        }
    }

    static int partition(int nums[], int l, int r) {
        int pivotIdx = l - 1;
        int pivot = nums[r];

        for (int i = l; i < r; i++) {
            if (nums[i] < pivot) {
                pivotIdx++;
                swap(pivotIdx, i, nums);
            }
        }

        swap(pivotIdx + 1, r, nums);
        return pivotIdx + 1;
    }

    static void swap(int a, int b, int nums[]) {
        int temp = nums[a];
        nums[a] = nums[b];
        nums[b] = temp;
    }
}
